### PR TITLE
Emotions on posts

### LINKED ISSUE #384: 🚀 [Feature] Emotion Reactions on Posts
### Emotion Reactions on Posts
image

- [ ] Emotions
- [ ] Image active + inactive @appinteractive can you provide the images?
- [ ] ~Percentage~ Absolute values by Userreaction distribution
- [ ] User Interaction (click/click again -> change)

> HC users wish better filter possibilities in order to suppress "annoying" user and usercontent from their personal Newsfeed. Extending the range of emojis from 5-10 is a good and easy way to create more comfort.

Emotions:
- Funny
- Happy
- Surprised
- Cry
- Angry
- Love (heart)
- Disconnected (broken heart)
- fear (bibbernd)
- questioning (questioning emoji)
- resignated (hand touching face)
- help! (siren on head feat. Minion :-D)  

Note: Once selected deselection is not possible

### Userstory
```
As a User
  I`d like posts to be tagged with emotions
  in order to be able to filter them by such
```

### Design & Layout - Alpha

No User Selection:
image

With User Selection:
image

### Design & Layout - Nitro
missing

### PR DESCRIPTION
**Pull request by mattwr18**
### Issues

- fixes #384

### CODE REVIEW

**Comment by appinteractive** [comment]
@mattwr18 I love the term „emoted“, you made my day. 😅 is it Even s Word?

**Comment by mattwr18** [comment]
That's great @appinteractive, I'm happy to have made your day :smile_cat: 
especially, when I didn't purposely set out to :stuck_out_tongue: 
yes, it's the past tense of emote :rofl:

**Comment by appinteractive** [comment]
Learned something again 😌

File: backend/src/schema/resolvers/posts.spec.js [hunk 1]
```diff
@@ -383,3 +385,220 @@ describe('DeletePost', () => {
...
+          data: 'cry',
+        })
+        await factory.emote({
+          from: authorParams.id,
+          to: 'p1376',
+          data: 'happy',
+        })
+      })
+
+      describe('but not the emoter', () => {
```
Comment (Tirokk) [inline backend/src/schema/resolvers/posts.spec.js]: I love the word `emoter `! 😍

File: webapp/locales/de.json [hunk 2]
```diff
@@ -415,6 +415,14 @@
...
     "categories": {
       "infoSelectedNoOfMaxCategories": "{chosen} von {max} Kategorien ausgewählt"
+    },
+    "emotions-label": {
+      "funny": "Lustig",
+      "happy": "Glücklich",
+      "surprised": "Erstaunt",
+      "cry": "Zum Weinen",
+      "angry": "Verärgert",
+      "emoted": "emotiert"
```
Comment (Tirokk) [inline webapp/locales/de.json]: The word `emotiert` just doesn’t exist in german, as far as I know. And I even don’t know anything similar.
I would set the word to `angegeben`.

File: webapp/pages/post/_id/_slug/index.spec.js [hunk 3]
```diff
@@ -31,7 +31,7 @@ describe('PostSlug', () => {
       $filters: {
         truncate: a => a,
       },
-      // If you mocking router, than don't use VueRouter with lacalVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
+      // If you are mocking the router, then don't use VueRouter with localVue: https://vue-test-utils.vuejs.org/guides/using-with-vue-router.html
```
Comment (Tirokk) [inline webapp/pages/post/_id/_slug/index.spec.js]: That’s an unrelated file !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
😂

File: webapp/pages/post/_id/_slug/index.vue [hunk 4]
```diff
@@ -199,4 +217,12 @@ export default {
     }
   }
 }
+@media only screen and (max-width: 960px) {
+  .shout-button {
+    float: left;
+  }
+  // .emotions-buttons-mobile {
+  //   margin-top: -80px;
+  // }
```
Comment (Tirokk) [inline webapp/pages/post/_id/_slug/index.vue]: May remove the comment?

File: backend/src/schema/types/type/Post.gql [hunk 5]
```diff
@@ -89,4 +91,10 @@ type Mutation {
     language: String
     categoryIds: [ID]
   ): Post
+  RemovePostEmotions(from: _UserInput!, to: _PostInput!, data: _EMOTEDInput!): Boolean!
+}
+
+type Query {
+  postsEmotionsCountByEmotion(postId: ID!, data: _EMOTEDInput!): Int!
```
Comment (roschaefer) [inline backend/src/schema/types/type/Post.gql]: ~I don't like STRANGECase, but I'm not going to request changes because of that - but you and everyone else are welcome to change it in the next PR :wink:~
Found improvements and would like to request changes - so you have the option to implement it in this PR :gift: 

Don't use STRANGECase, pleASE

File: backend/src/middleware/permissionsMiddleware.js [hunk 6]
```diff
@@ -27,6 +27,12 @@ const onlyYourself = rule({
   return context.user.id === args.id
 })
 
+const isMyEmotion = rule({
+  cache: 'no_cache',
+})(async (parent, args, context, info) => {
+  return context.user.id === args.from.id
```
Comment (roschaefer) [inline backend/src/middleware/permissionsMiddleware.js]: Eww - implement your own custom resolvers and you don't have to check that anymore.

eww

File: backend/src/schema/resolvers/posts.js [hunk 7]
```diff
@@ -70,5 +71,76 @@ export default {
...
+          from: { id: record.get('userFrom').properties.id },
+          to: { id: record.get('postTo').properties.id },
+          emotion: record.get('emotedRelation').properties.emotion,
+        }
+      })
+      return emoted
+    },
+    RemovePostEmotions: async (object, params, context, resolveInfo) => {
+      const session = context.driver.session()
+      const { from, to, data } = params
```
Comment (roschaefer) [inline backend/src/schema/resolvers/posts.js]:
```suggestion
      const { to, data } = params
      const { id: from } = context.user
```

File: backend/src/seed/factories/index.js [hunk 8]
```diff
@@ -128,6 +128,23 @@ export default function Factory(options = {}) {
       this.lastResponse = await cleanDatabase({ driver: this.neo4jDriver })
       return this
     },
+    async emote({ from, to, data }) {
```
Comment (roschaefer) [inline backend/src/seed/factories/index.js]: If you use `neode` you wouldn't have to make a HTTP request over the wire.

File: webapp/components/EmotionsButtons/EmotionsButtons.vue [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      currentUser: 'auth/user',
+    }),
+  },
+  created() {
+    Object.keys(this.postsEmotionsCountByEmotion).map(emotion => {
+      this.emotionsCount(emotion)
+    })
+  },
+  methods: {
+    iconPath(emotion) {
```
Comment (roschaefer) [inline webapp/components/EmotionsButtons/EmotionsButtons.vue]: That's a computed property!
...
Well I just saw you're using the plural `EmotionButtons`. Why not `EmotionButton`? Tests are guiding you how you have to set the boundaries :heart: tests.

File: webapp/components/EmotionsButtons/EmotionsButtons.vue [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+      if (index > -1) {
+        return true
+      }
+      return false
+    },
+    emotionsCount(emotion) {
+      this.$apollo
+        .query({
+          query: gql`
+            query($postId: ID!, $data: _EMOTEDInput!) {
```
Comment (roschaefer) [inline webapp/components/EmotionsButtons/EmotionsButtons.vue]: If you want to query the data in one go, just have`Emotions` doing the API hard work and `EmotionButton` doing the visual stuff. Use `this.$emit` to learn what button was pushed.

File: webapp/pages/post/_id/_slug/index.spec.js (hunk 3 above)
Comment (roschaefer) [reply webapp/pages/post/_id/_slug/index.spec.js]: !!!111eleven :angry:

Comment (mattwr18) [reply webapp/pages/post/_id/_slug/index.spec.js]: what is this supposed to mean? hahaha

File: backend/src/schema/types/type/Post.gql (hunk 5 above)
Comment (mattwr18) [reply backend/src/schema/types/type/Post.gql]: what's STRANGECase? I found this online https://www.artrix.co.uk/whats-on/all-shows/the-strange-case-of-dr-jekyll-and-mr-hyde

the case used in this example is called camel case and is used quite a lot in javascript, do you prefer pascal case?
https://medium.com/better-programming/string-case-styles-camel-pascal-snake-and-kebab-case-981407998841

File: backend/src/middleware/permissionsMiddleware.js (hunk 6 above)
Comment (mattwr18) [reply backend/src/middleware/permissionsMiddleware.js]: good catch

File: backend/src/schema/resolvers/posts.js (hunk 7 above)
Comment (mattwr18) [reply backend/src/schema/resolvers/posts.js]: didn't think of that, but like it!!

File: webapp/pages/post/_id/_slug/index.spec.js (hunk 3 above)
Comment (roschaefer) [reply webapp/pages/post/_id/_slug/index.spec.js]: When people get really angry, they use a lot of exclamation marks and don't manage to keep the shift key pressed!!!11
